File: implementations/python/exchange_arguments.py
"""
Greedy Exchange Arguments Verification in Python.

Implements Interval Scheduling (Earliest Finish Time) and Minimizing Lateness (EDD)
with brute-force differential testing proving greedy optimality via the exchange property.
"""

import itertools
import unittest
from typing import List, NamedTuple


class Interval(NamedTuple):
    start: int
    finish: int
    id: int


def interval_scheduling_greedy(intervals: List[Interval]) -> List[Interval]:
    sorted_ivs = sorted(intervals, key=lambda x: (x.finish, x.start))
    selected: List[Interval] = []
    last_finish = -1

    for iv in sorted_ivs:
        if iv.start >= last_finish:
            selected.append(iv)
            last_finish = iv.finish

    return selected


def interval_scheduling_brute_force(intervals: List[Interval]) -> int:
    n = len(intervals)
    max_len = 0

    for r in range(1, n + 1):
        for subset in itertools.combinations(intervals, r):
            sorted_sub = sorted(subset, key=lambda x: x.start)
            compatible = all(
                sorted_sub[i].start >= sorted_sub[i - 1].finish
                for i in range(1, len(sorted_sub))
            )
            if compatible:
                max_len = max(max_len, len(sorted_sub))

    return max_len


class Job(NamedTuple):
    duration: int
    deadline: int
    id: int


def minimize_lateness_edd(jobs: List[Job]) -> int:
    sorted_jobs = sorted(jobs, key=lambda x: x.deadline)
    curr_time = 0
    max_l = 0
    for j in sorted_jobs:
        curr_time += j.duration
        lateness = max(0, curr_time - j.deadline)
        max_l = max(max_l, lateness)
    return max_l


def minimize_lateness_brute_force(jobs: List[Job]) -> int:
    min_l = float("inf")
    for p in itertools.permutations(jobs):
        curr_time = 0
        max_l = 0
        for j in p:
            curr_time += j.duration
            l = max(0, curr_time - j.deadline)
            max_l = max(max_l, l)
        min_l = min(min_l, max_l)
    return int(min_l)


class TestExchangeArguments(unittest.TestCase):
    def test_interval_scheduling_optimality(self):
        intervals = [
            Interval(1, 4, 1), Interval(3, 5, 2), Interval(0, 6, 3),
            Interval(5, 7, 4), Interval(3, 9, 5), Interval(5, 9, 6),
            Interval(6, 10, 7), Interval(8, 11, 8), Interval(8, 12, 9),
            Interval(2, 14, 10), Interval(12, 16, 11)
        ]
        greedy = interval_scheduling_greedy(intervals)
        brute_cardinality = interval_scheduling_brute_force(intervals)
        self.assertEqual(len(greedy), brute_cardinality)
        self.assertEqual(len(greedy), 4)

    def test_minimizing_lateness_optimality(self):
        jobs = [
            Job(3, 6, 1),
            Job(2, 8, 2),
            Job(1, 9, 3),
            Job(4, 9, 4),
            Job(3, 14, 5),
            Job(2, 15, 6)
        ]
        edd = minimize_lateness_edd(jobs)
        brute = minimize_lateness_brute_force(jobs)
        self.assertEqual(edd, brute)


if __name__ == "__main__":
    unittest.main()

File: implementations/cpp/exchange_arguments.cpp
/**
 * @file exchange_arguments.cpp
 * @brief Algorithmic validation of Greedy Exchange Arguments.
 *
 * Implements Interval Scheduling (Earliest Finish Time) and Minimizing Lateness (Earliest Due Date)
 * with differential brute-force verification proving greedy optimality via the exchange property.
 * Verified with C++17 (-std=c++17 -O3 -Wall -Wextra -Werror).
 */

#include <iostream>
#include <vector>
#include <cstdint>
#include <cassert>
#include <algorithm>
#include <numeric>

namespace dsa {

struct Interval {
    int start;
    int finish;
    int id;
};

/**
 * @brief Greedy Interval Scheduling: Selects maximum non-overlapping intervals by Earliest Finish Time.
 */
inline std::vector<Interval> interval_scheduling_greedy(std::vector<Interval> intervals) {
    if (intervals.empty()) return {};

    std::sort(intervals.begin(), intervals.end(), [](const Interval& a, const Interval& b) {
        if (a.finish != b.finish) return a.finish < b.finish;
        return a.start < b.start;
    });

    std::vector<Interval> selected;
    int last_finish = -1;

    for (const auto& iv : intervals) {
        if (iv.start >= last_finish) {
            selected.push_back(iv);
            last_finish = iv.finish;
        }
    }
    return selected;
}

/**
 * @brief Brute-force Oracle for Interval Scheduling: Explores all 2^N subsets to find true maximum.
 */
inline size_t interval_scheduling_brute_force(const std::vector<Interval>& intervals) {
    size_t n = intervals.size();
    size_t max_count = 0;

    for (uint32_t mask = 0; mask < (1U << n); ++mask) {
        std::vector<Interval> subset;
        for (size_t i = 0; i < n; ++i) {
            if (mask & (1U << i)) {
                subset.push_back(intervals[i]);
            }
        }

        // Check compatibility
        std::sort(subset.begin(), subset.end(), [](const Interval& a, const Interval& b) {
            return a.start < b.start;
        });

        bool compatible = true;
        for (size_t i = 1; i < subset.size(); ++i) {
            if (subset[i].start < subset[i - 1].finish) {
                compatible = false;
                break;
            }
        }

        if (compatible) {
            max_count = std::max(max_count, subset.size());
        }
    }
    return max_count;
}

// --- Minimizing Lateness Scheduling ---

struct Job {
    int duration;
    int deadline;
    int id;
};

/**
 * @brief Earliest Due Date (EDD) Minimizing Lateness.
 * Returns maximum lateness L = max(0, finish - deadline).
 */
inline int minimize_lateness_edd(std::vector<Job> jobs) {
    std::sort(jobs.begin(), jobs.end(), [](const Job& a, const Job& b) {
        return a.deadline < b.deadline;
    });

    int current_time = 0;
    int max_lateness = 0;

    for (const auto& j : jobs) {
        current_time += j.duration;
        int lateness = std::max(0, current_time - j.deadline);
        max_lateness = std::max(max_lateness, lateness);
    }
    return max_lateness;
}

/**
 * @brief Brute-force all N! permutations to verify EDD optimality.
 */
inline int minimize_lateness_brute_force(std::vector<Job> jobs) {
    std::vector<size_t> p(jobs.size());
    std::iota(p.begin(), p.end(), 0);

    int best_lateness = 1e9;

    do {
        int current_time = 0;
        int max_l = 0;
        for (size_t idx : p) {
            current_time += jobs[idx].duration;
            int lateness = std::max(0, current_time - jobs[idx].deadline);
            max_l = std::max(max_l, lateness);
        }
        best_lateness = std::min(best_lateness, max_l);
    } while (std::next_permutation(p.begin(), p.end()));

    return best_lateness;
}

} // namespace dsa

int main() {
    std::cout << "Running Greedy Exchange Arguments verification..." << std::endl;

    // 1. Interval Scheduling: Test against Brute Force across diverse inputs
    std::vector<dsa::Interval> test_intervals = {
        {1, 4, 1}, {3, 5, 2}, {0, 6, 3}, {5, 7, 4}, {3, 9, 5},
        {5, 9, 6}, {6, 10, 7}, {8, 11, 8}, {8, 12, 9}, {2, 14, 10}, {12, 16, 11}
    };

    auto greedy_res = dsa::interval_scheduling_greedy(test_intervals);
    size_t brute_count = dsa::interval_scheduling_brute_force(test_intervals);

    assert(greedy_res.size() == brute_count);
    assert(greedy_res.size() == 4); // Standard known optimal cardinality for this classic set

    // 2. Minimizing Lateness: Test EDD optimality vs N! permutations
    std::vector<dsa::Job> test_jobs = {
        {3, 6, 1},
        {2, 8, 2},
        {1, 9, 3},
        {4, 9, 4},
        {3, 14, 5},
        {2, 15, 6}
    };

    int edd_lateness = dsa::minimize_lateness_edd(test_jobs);
    int optimal_lateness = dsa::minimize_lateness_brute_force(test_jobs);

    assert(edd_lateness == optimal_lateness);

    std::cout << "[PASS] Greedy Interval Scheduling EFT matched brute-force optimal cardinality (" << brute_count << ")." << std::endl;
    std::cout << "[PASS] Earliest Due Date scheduling matched N! permutation optimal lateness (" << optimal_lateness << ")." << std::endl;
    std::cout << "All Greedy Exchange Arguments assertions passed successfully!" << std::endl;
    return 0;
}
